src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs

```diff
@@ -14,211 +24,3020 @@ namespace Microsoft.CodeAnalysis.CSharp
+        /// is true for other
+        /// constructs that can cause an exception to be raised such as a throw statement or method
+        /// invocation.
+        /// The idea is that we don't know if the branch will eventually reach its destination
+        /// because of an intervening finally block that cannot complete normally.  So we store them
+        /// up and handle them as we complete processing each construct.  At the end of a block, if
+        /// there are any pending branches to a label in that block we process the branch.  Otherwise
+        /// we relay it up to the enclosing construct as a pending branch of the enclosing
+        /// construct.
+        /// </summary>
+        internal class PendingBranch
         {
-            if (_trackUnassignments)
+            public readonly BoundNode Branch;
+            public TLocalState State;
+            public LabelSymbol Label
             {
-                Optional<TLocalState> oldTryState = _tryState;
-                _tryState = AllBitsSet();
-                VisitTryBlock(tryBlock, node, ref tryState);
-                var tempTryStateValue = _tryState.Value;
-                IntersectWith(ref tryState, ref tempTryStateValue);
-                if (oldTryState.HasValue)
+                get
                 {
-                    var oldTryStateValue = oldTryState.Value;
-                    IntersectWith(ref oldTryStateValue, ref tempTryStateValue);
-                    oldTryState = oldTryStateValue;
+                    if (Branch == null) return null;
+                    switch (Branch.Kind)
+                    {
+                        case BoundKind.GotoStatement: return ((BoundGotoStatement)Branch).Label;
+                        case BoundKind.ConditionalGoto: return ((BoundConditionalGoto)Branch).Label;
+                        case BoundKind.BreakStatement: return ((BoundBreakStatement)Branch).Label;
+                        case BoundKind.ContinueStatement: return ((BoundContinueStatement)Branch).Label;
+                        case BoundKind.PatternSwitchLabel: return ((BoundPatternSwitchLabel)Branch).Label;
+                        default: return null;
+                    }
                 }
+            }
 
-                _tryState = oldTryState;
+            public PendingBranch(BoundNode branch, TLocalState state)
+            {
+                this.Branch = branch;
+                this.State = state.Clone();
             }
-            else
+        }
+
+        /// <summary>
+        /// Perform a single pass of flow analysis.  Note that after this pass,
+        /// this.backwardBranchChanged indicates if a further pass is required.
+        /// </summary>
+        protected virtual ImmutableArray<PendingBranch> Scan(ref bool badRegion)
+        {
+            var oldPending = SavePending();
+            Visit(methodMainNode);
+            this.Unsplit();
+            RestorePending(oldPending);
```

Comment:
could you doc the logic of this method.  It seems rather important, but it isn't clear to me why it needs to do thins like 'Unsplit'